pkg/lease/lease_test.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lease
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"sort"
+	"sync"
+	"testing"
+	"time"
+
+	monitoring "cloud.google.com/go/monitoring/apiv3/v2"
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"google.golang.org/api/option"
+	monitoredres_pb "google.golang.org/genproto/googleapis/api/monitoredres"
+	monitoring_pb "google.golang.org/genproto/googleapis/monitoring/v3"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/test/bufconn"
+	empty_pb "google.golang.org/protobuf/types/known/emptypb"
+)
+
+type testMetricService struct {
+	monitoring_pb.MetricServiceServer // Inherit all interface methods
+
+	latency        time.Duration
+	frequencyLimit time.Duration
+	// TODO(freinartz): add error probability
+
+	mtx          sync.Mutex
+	lastInterval *monitoring_pb.TimeInterval
+}
+
+func (srv *testMetricService) CreateTimeSeries(ctx context.Context, req *monitoring_pb.CreateTimeSeriesRequest) (*empty_pb.Empty, error) {
+	select {
+	case <-time.After(srv.latency):
+	case <-ctx.Done():
+		return nil, status.Error(codes.Canceled, "canceled")
+	}
+	srv.mtx.Lock()
+	defer srv.mtx.Unlock()
+
+	// Assume well-formed requests and just validate timestamps according to the logic
+	// of the real API.
+	// We expect all clients to write against the same lease for the test.
+	iv := req.TimeSeries[0].Points[0].Interval
+	// First write always succeeds.
+	if srv.lastInterval != nil {
+		// The end timestamp must not go into the past and not be too shortly after
+		// the last one.
+		if iv.EndTime.AsTime().Before(srv.lastInterval.EndTime.AsTime()) {
```

Comment:
IIUC these error conditions are tolerated by the lease where the state machine will proceed to follow mode. Is there utility of the error message flavors? AFAICT if they are triggered, they wouldn't cause test failures.

Reply:
The implementation only checks the code so the message here is just for potential debugging.